### PR TITLE
Update SuggestedAction to track successful application

### LINKED ISSUE #34165: CSharpUpdateProjectToAllowUnsafe.CPSProject_GeneralPropertyGroupUpdated Fails
[Test Logs](https://dev.azure.com/dnceng/public/_build/results?buildId=125135&view=ms.vss-test-web.build-test-results-tab)

```
Assert.Contains() Failure
Not found: (filter expression)
In value: WhereEnumerableIterator<XElement> [<PropertyGroup>
 <TargetFramework>netstandard2.0</TargetFramework>
</PropertyGroup>]
```

### LINKED ISSUE #35909: Integration test CPSProject_GeneralPropertyGroupUpdated failed in CI
https://dev.azure.com/dnceng/public/_build/results?buildId=197521&view=ms.vss-test-web.build-test-results-tab&runId=4796290&paneView=debug&resultId=100389

```
Assert.Contains() Failure\r\nNot found: (filter expression)\r\nIn value: WhereEnumerableIterator<XElement> [<PropertyGroup>\r\n <TargetFramework>netstandard2.0</TargetFramework>\r\n</PropertyGroup>]

Stack trace
   at Roslyn.VisualStudio.IntegrationTests.AbstractUpdateProjectTest.VerifyPropertyOutsideConfiguration(XElement projectElement, String name, String value) in /_/src/VisualStudio/IntegrationTest/IntegrationTests/AbstractUpdateProjectTest.cs:line 31
   at Roslyn.VisualStudio.IntegrationTests.CSharp.CSharpUpdateProjectToAllowUnsafe.CPSProject_GeneralPropertyGroupUpdated() in /_/src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpUpdateProjectToAllowUnsafe.cs:line 44
Build 20190522.24 Failed

Showing filters 1 through 3
```

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #35909 by silently retrying to apply a code fix if the application failed due to concurrent changes in the workspace.

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActions/SuggestedAction.cs [hunk 1]
```diff
@@ -308,5 +312,18 @@ public override int GetHashCode()
...
+            => new TestAccessor(this);
+
+        internal readonly struct TestAccessor
+        {
+            private readonly SuggestedAction _suggestedAction;
+
+            public TestAccessor(SuggestedAction suggestedAction)
+                => _suggestedAction = suggestedAction;
+
+            public ref bool IsApplied => ref _suggestedAction._isApplied;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActions/SuggestedAction.cs]: is this acutally set in tests?

File: src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs [hunk 2]
```diff
@@ -257,10 +260,11 @@ internal class CodeActionEditHandlerService : ForegroundThreadAffinitizedObject,
             }
         }
 
-        private static void ProcessOperations(
+        private static bool ProcessOperations(
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs]: maybe document the contract.

File: src/VisualStudio/IntegrationTest/TestUtilities/InProcess/TextViewWindow_InProc.cs [hunk 3]
```diff
@@ -329,24 +329,28 @@ private async Task<IEnumerable<ISuggestedAction>> GetLightBulbActionsAsync(ILigh
...
                 var activeTextView = GetActiveTextView();
-                await lightBulbAction(activeTextView);
+                return await lightBulbAction(activeTextView);
             });
 
             if (blockUntilComplete)
             {
-                task.Join();
+                var result = task.Join();
+                DismissLightBulbSession();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/IntegrationTest/TestUtilities/InProcess/TextViewWindow_InProc.cs]: why this change?

File: src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/TextViewWindow_OutOfProc.Verifier.cs [hunk 4]
```diff
@@ -30,12 +31,21 @@ public Verifier(TTextViewWindow textViewWindow, VisualStudioInstance instance)
...
-                CodeActions(expectedItems, applyFix ? expectedItem : null, verifyNotShowing,
-                    ensureExpectedItemsAreOrdered, fixAllScope, blockUntilComplete);
+
+                bool applied;
+                do
+                {
+                    cancellationTokenSource.Token.ThrowIfCancellationRequested();
+
+                    applied = CodeActions(expectedItems, applyFix ? expectedItem : null, verifyNotShowing,
+                        ensureExpectedItemsAreOrdered, fixAllScope, blockUntilComplete) ?? true;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/TextViewWindow_OutOfProc.Verifier.cs]: can you explain the reason you want the `??` true here?

File: src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/TextViewWindow_OutOfProc.Verifier.cs [hunk 5]
```diff
@@ -30,12 +31,21 @@ public Verifier(TTextViewWindow textViewWindow, VisualStudioInstance instance)
...
+                } while (!applied && applyFix);
             }
 
-            public void CodeActions(
+            public bool? CodeActions(
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/TextViewWindow_OutOfProc.Verifier.cs]: doc.

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActions/SuggestedAction.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActions/SuggestedAction.cs]: no one seems to assign.  i would just expose as bool.

File: src/VisualStudio/IntegrationTest/TestUtilities/InProcess/TextViewWindow_InProc.cs (hunk 3 above)
Comment (sharwell) [reply src/VisualStudio/IntegrationTest/TestUtilities/InProcess/TextViewWindow_InProc.cs]: Code fix suggestions are not recomputed while the light bulb is visible. Either way, after applying a code fix, we need to dismiss the session to match the behavior you get when a user applies a code fix by interacting with the menus.

Comment (CyrusNajmabadi) [reply src/VisualStudio/IntegrationTest/TestUtilities/InProcess/TextViewWindow_InProc.cs]: ok.  so this was a bug, and we shoudl have done this regardless of tracking if things were applied or not?

**Comment by CyrusNajmabadi** [comment]
Looks reasonable to me.  A little bit of docs woudl help out.  Thanks!

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActions/SuggestedAction.cs (hunk 1 above)
Comment (sharwell) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActions/SuggestedAction.cs]: This form means the tests operate as though they directly have access to the backing field. The backing field is mutable.

File: src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs (hunk 2 above)
Comment (sharwell) [reply src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs]: updated

File: src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/TextViewWindow_OutOfProc.Verifier.cs (hunk 4 above)
Comment (sharwell) [reply src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/TextViewWindow_OutOfProc.Verifier.cs]: No longer using `??`

File: src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/TextViewWindow_OutOfProc.Verifier.cs (hunk 5 above)
Comment (sharwell) [reply src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/TextViewWindow_OutOfProc.Verifier.cs]: Updated

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActions/SuggestedAction.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActions/SuggestedAction.cs]: I'm on hte fence.  I get the desire here.  But i also think it's valuable if hte test accessor (or, frankly, any consumer) can have minimal access.  But if you feel this is really a core part of the design, i'm fine with you preserving this.

**Comment by heejaechang** [comment]
looks good for testing.

if we make it to actually return reasoning why it is failed, we could actually use it in real product I believe?

**Comment by sharwell** [comment]
Merging with failed integration tests because it has a demonstrated improvement on integration test pass rates.

**Comment by heejaechang** [comment]
I am not talking about retrying in the test, I meant when TryApplyChange failed (due to for example, Workspace.TryApplyChange return false due to workspace being changed underneath for example due to file change notification), right now, we jsut silently fail. there is no indication to users that we have failed.

we could make product to actually show some kind of UI telling users it has failed due to X reason.

**Comment by sharwell** [comment]
@heejaechang Yes that's the way I understood your comment. I agree that this is an unfortunate limitation of the current UI and it would be nice to improve it.
